samza-core/src/main/java/org/apache/samza/task/TaskFactoryUtil.java

```diff
@@ -39,28 +39,34 @@
   private static final Logger log = LoggerFactory.getLogger(TaskFactoryUtil.class);
 
   /**
-   * This method loads a task factory class based on the configuration
+   * This method creates a task factory class based on the configuration and {@link StreamApplication}
    *
    * @param config  the {@link Config} for this job
    * @param runner  the {@link ApplicationRunner} to run this job
    * @return  a task factory object, either a instance of {@link StreamTaskFactory} or {@link AsyncStreamTaskFactory}
    */
-  public static Object fromTaskClassConfig(Config config, ApplicationRunner runner) {
+  public static Object createTaskFactory(Config config, StreamApplication streamApp, ApplicationRunner runner) {
+    return (streamApp != null) ? createStreamOperatorTaskFactory(streamApp, runner) : fromTaskClassConfig(config);
+  }
 
-    String taskClassName;
+  private static StreamTaskFactory createStreamOperatorTaskFactory(StreamApplication streamApp, ApplicationRunner runner) {
+    return () -> new StreamOperatorTask(streamApp, runner);
+  }
 
+  /**
+   * Create {@link StreamTaskFactory} or {@link AsyncStreamTaskFactory} based on the configured task.class.
+   * @param config
+   * @return task factory instance
+   */
+  public static Object fromTaskClassConfig(Config config) {
```

Comment:
Since the goal of adding this class was to clean up task factory creation, I'd prefer to consolidate all the logic and have #createTaskFactory be the only public method. Maybe you can just pass it null for streamApp to get the same behavior in tests?